Make factory methods for labels and relationship types static methods on the interfaces

thobe opened the pull request:
With Java 8 we can have static methods on interfaces, so there is no need for DynamicLabel and DynamicRelationshipType. Instead we can move the corresponding factory methods to the Label and RelationshipType interfaces respectively.

chrisvest commented on community/kernel/src/main/java/org/neo4j/graphdb/Label.java:
```diff
@@ -68,4 +67,26 @@
+                return name;
+            }
+        };
+    }
```
I think it would be good to implement `equals` and `hashCode` on these guys, since I suspect people will naturally think of them as value objects.

chrisvest replied:
Okay, I see that this is very carefully documented as a terrible idea.

chrisvest commented on community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/infrastructure/ValueRelationship.java:
```diff
@@ -71,7 +70,8 @@ public static ValueRelationship unpackFields( Neo4jPack.Unpacker unpacker )
 
-        RelationshipType relType = DynamicRelationshipType.withName( relTypeName );
+        //        return new DynamicRelationshipType( name );
+        RelationshipType relType = RelationshipType.withName( relTypeName );
```
Why this comment?

chrisvest commented on community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/infrastructure/ValueUnboundRelationship.java:
```diff
@@ -60,7 +59,8 @@ public static ValueUnboundRelationship unpackFields( Neo4jPack.Unpacker unpacker
 
-        RelationshipType relType = DynamicRelationshipType.withName( relTypeName );
+        //        return new DynamicRelationshipType( name );
+        RelationshipType relType = RelationshipType.withName( relTypeName );
```
And this comment?

chrisvest commented on community/bolt/src/test/java/org/neo4j/bolt/v1/docs/BoltValueDocTest.java:
```diff
@@ -171,11 +171,12 @@ else if ( type.equalsIgnoreCase( "map" ) )
         {
-            return new ValueRelationship( 12, 1, 2, DynamicRelationshipType.withName( "KNOWS" ), map() );
+            //        return new DynamicRelationshipType( name );
+            return new ValueRelationship( 12, 1, 2, RelationshipType.withName( "KNOWS" ), map() );
```
And this one?

chrisvest commented on community/bolt/src/test/java/org/neo4j/bolt/v1/messaging/MessageFormatTest.java:
```diff
@@ -122,14 +122,15 @@ public void shouldSerializeBasicTypes() throws Throwable
-        assertSerializesNeoValue( new ValueRelationship( 12l, 1l, 2l, DynamicRelationshipType.withName( "KNOWS" ),
+        //        return new DynamicRelationshipType( name );
+        assertSerializesNeoValue( new ValueRelationship( 12l, 1l, 2l, RelationshipType.withName( "KNOWS" ),
                 map( "name", "Bob", "age", 14 ) ) );
```
And this one too?

chrisvest commented on community/bolt/src/test/java/org/neo4j/bolt/v1/messaging/example/Nodes.java:
```diff
@@ -34,15 +34,33 @@
+                asList( Label.label( "Person" ), Label.label( "Employee" ) ),
+                map( "name", "Bob", "age", 44L ) );
+    }
+
```
You cannot make these assignments in `static { … }` clauses while also having final field assignments at the field definitions.

chrisvest commented on community/consistency-check-legacy/src/test/java/org/neo4j/legacy/consistency/ConsistencyCheckServiceIntegrationTest.java:
```diff
@@ -207,7 +206,8 @@ protected void generateInitialData( GraphDatabaseService graphDb )
                 Node node2 = set( graphDb.createNode(), property( "key", "value" ) );
-                node1.createRelationshipTo( node2, DynamicRelationshipType.withName( "C" ) );
+                //        return new DynamicRelationshipType( name );
+                node1.createRelationshipTo( node2, RelationshipType.withName( "C" ) );
```
This comment.

chrisvest commented on community/consistency-check-legacy/src/test/java/org/neo4j/legacy/consistency/checking/full/ExecutionOrderIntegrationTest.java:
```diff
@@ -90,7 +90,8 @@ protected void generateInitialData( GraphDatabaseService graphDb )
                 Node node2 = set( graphDb.createNode(), property( "key", "value" ) );
-                node1.createRelationshipTo( node2, DynamicRelationshipType.withName( "C" ) );
+                //        return new DynamicRelationshipType( name );
+                node1.createRelationshipTo( node2, RelationshipType.withName( "C" ) );
```
Apparently you have a lot of these comments.

chrisvest commented on community/embedded-examples/src/main/java/org/neo4j/examples/orderedpath/OrderedPath.java:
```diff
@@ -34,12 +34,11 @@
-            REL3 = withName( "REL3" );
+    private static final RelationshipType REL1 = RelationshipType.withName( "REL1" );
+    private static final RelationshipType REL2 = RelationshipType.withName( "REL2" );
+    private static final RelationshipType REL3 = RelationshipType.withName( "REL3" );
```
Good.

chrisvest commented on community/graph-matching/src/test/java/matching/TestMatchingOfCircularPattern.java:
```diff
@@ -98,9 +100,12 @@ public static void setupGraph()
+        //        return new DynamicRelationshipType( name );
+        user1.createRelationshipTo( user3, RelationshipType.withName( "FOLLOWS" ) );
+        //        return new DynamicRelationshipType( name );
+        user.createRelationshipTo( user2, RelationshipType.withName( "FOLLOWS" ) );
```
I suspect they are coming from a method you inlined to fix all calls to `DynamicRelationshipType.withName()`.

chrisvest commented on community/graph-matching/src/test/java/matching/TestMatchingOfCircularPattern.java:
```diff
@@ -224,6 +234,8 @@ public void verifyCount()
     {
+        //        return new DynamicRelationshipType( name );
+        //        return new DynamicRelationshipType( name );
         return startNode.traverse( Order.BREADTH_FIRST, stopAtDepth( 2 ),
```
Imagine the testing the JetBrains guys must do, to make this comment placement work reliably.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/core/NodeProxy.java:
```diff
@@ -812,7 +810,8 @@ public RelationshipType apply( int id )
                 {
-                    return DynamicRelationshipType.withName( statement.readOperations().relationshipTypeGetName( id ) );
+                    return RelationshipType.withName( statement.readOperations().relationshipTypeGetName( id ) );
+//        return new DynamicRelationshipType( name );
```
This one they fucked up, though.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/counts/CompositeCountsTest.java:
```diff
@@ -50,31 +47,48 @@ public void shouldReportNumberOfRelationshipsFromNodesWithGivenLabel() throws Ex
+        //        return new DynamicRelationshipType( name );
+        numberOfRelationshipsMatching( null, RelationshipType.withName( "BETA" ), Label.label( "Bar" ) ).shouldBe( 2 );
+        //        return new DynamicRelationshipType( name );
+        numberOfRelationshipsMatching( null, RelationshipType.withName( "GAMMA" ), Label.label( "Bar" ) ).shouldBe( 0 );
```
Oh dear.

chrisvest commented on community/kernel/src/main/java/org/neo4j/graphdb/Label.java:
```diff
@@ -68,4 +67,26 @@
      */
     String name();
+
+    static Label label( String name )
```
Needs some javadoc.

chrisvest commented on community/kernel/src/main/java/org/neo4j/graphdb/RelationshipType.java:
```diff
@@ -98,4 +100,26 @@
      */
     String name();
+
+    static RelationshipType withName( String name )
```
And some javadoc here would be good as well.

chrisvest commented:
@thobe I've done the first review round. Looks like there's still some work to do.

thobe commented:
@chrisvest yes, it does look like the inlining didn't work very well...

On the topic of `equals` and `hashCode`, the previous implementation in `DynamicLabel` was not symmetric. In fact it was impossible to make symmetric with the enum pattern. But we could make an implementation that makes instances of the "dynamic" type equal to other instances of the same class (iff the name is equal).

thobe committed:
Move factory methods to interface

With Java 8 we can have static methods on interfaces, so there is no
need for DynamicLabel and DynamicRelationshipType. Instead we can move
the corresponding factory methods to the Label and RelationshipType
interfaces respectively.

chrisvest commented:
@thobe There are test failures.
